### PR TITLE
feat(client/windows): introduce go plugin to call `fetchResource` as a library

### PR DESCRIPTION
**Pull request by jyyi1**
This PR introduces a new mechanism for Electron's renderer process to directly call Go functions.

It shows how this works by migrating the `fetchResource` function as an example.

### CODE REVIEW

File: client/go/outline/electron/go_plugin.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+package main
+
+/*
+#include <stdlib.h>  // for C.free
+
+// InvokeGoAPIResult is a struct used to pass result from Go to TypeScript boundary.
+typedef struct InvokeGoAPIResult_t
+{
+  // A string representing the result of the Go function call.
```
[the remark this answers is not in the document]
Comment (jyyi1) [reply client/go/outline/electron/go_plugin.go:23]: What do you mean here? It is aligned with the `const char *Output` field.

[the remark this answers is not in the document]
Comment (sbruens) [reply client/go/outline/electron/go_plugin.go:23]: It's not aligned for me:

Screenshot 2024-11-14 at 3 53 28 PM

Not a big deal though. Might be a GitHub whitespace rendering issue?

[the remark this answers is not in the document]
Comment (jyyi1) [reply client/go/outline/electron/go_plugin.go:23]: Got it! Spaces vs tabs! 😄

File: client/electron/app_paths.ts [hunk 2]
```diff
@@ -56,6 +56,17 @@ export function pathToEmbeddedTun2socksBinary() {
   );
 }
 
+export function pathToBackendLibrary() {
+  return path.join(
+    unpackedAppPath(),
+    'client',
+    'output',
+    'build',
+    isWindows ? 'windows' : 'linux',
```
[the remark this answers is not in the document]
Comment (jyyi1) [reply client/electron/app_paths.ts]: Done

File: client/go/outline/electron/go_plugin.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jyyi1) [reply client/go/outline/electron/go_plugin.go]: Fixed

File: client/go/outline/electron/go_plugin.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+/*
+#include <stdlib.h>  // for C.free
+
+// InvokeGoAPIResult is a struct used to pass result from Go to TypeScript boundary.
+typedef struct InvokeGoAPIResult_t
```
Comment (fortuna) [inline client/go/outline/electron/go_plugin.go:21]: I think you can define this struct in Go instead and only have the include here.

**Review by fortuna — APPROVED** [review]
Strong change. I love it!

File: client/go/outline/electron/go_plugin.go (hunk 3 above)
Comment (jyyi1) [reply client/go/outline/electron/go_plugin.go:21]: Do you mean define it as the `struct` as:

```go
type InvokeGoAPIResult struct {
  Output, ErrorJson *C.char
}
```

I'm afraid this is [not supported by CGo](https://pkg.go.dev/cmd/cgo#hdr-C_references_to_Go):

> Not all Go types can be mapped to C types in a useful way. Go struct types are not supported; use a C struct type. Go array types are not supported; use a C pointer.

**Pull request merged**